src/chunk new.rs

```diff
@@ -0,0 +1,289 @@
+                        .field("type", data)
+                        .field("critical", &is_critical(*data))
+                        .field("private", &is_private(*data))
+                        .field("reserved", &reserved_set(*data))
+                        .field("safecopy", &safe_to_copy(*data))
+                        .finish()
+                }
+            }
+        }
+
+        $(
+            $(#[doc=$desc])?
+            #[derive(PartialEq, Eq)]
+            pub struct $chunk;
+
+            impl Chunk for $chunk {
+                const TYPE: [u8; 4] = $type;
+                const MAX_LENGTH: usize = $size as usize;
+            }
+
+            impl std::ops::Deref for $chunk {
+                type Target = [u8; 4];
+
+                fn deref(&self) -> &Self::Target {
+                    &Self::TYPE
+                }
+            }
+
+            impl std::fmt::Debug for $chunk {
+                fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+                    f.debug_struct("ChunkType")
+                        .field("type", &stringify!($chunk))
+                        $(.field("description", &$desc))?
+                        .field("critical", &Self::is_critical())
+                        .field("private", &Self::is_private())
+                        .field("reserved", &Self::reserved_set())
+                        .field("safecopy", &Self::safe_to_copy())
+                        .finish()
+                }
+            }
+        )*
+    };
+}
+
+// -- Chunk type determination --
+
+/// Returns true if the chunk is critical.
+const fn is_critical(chunk: [u8; 4]) -> bool {
+    chunk[0] & 0b_00100000 == 0
+}
+
+/// Returns true if the chunk is private.
+const fn is_private(chunk: [u8; 4]) -> bool {
+    chunk[1] & 0b_00100000 != 0
+}
+
+/// Checks whether the reserved bit of the chunk name is set.
+/// If it is set the chunk name is invalid.
+const fn reserved_set(chunk: [u8; 4]) -> bool {
+    chunk[2] & 0b_00100000 != 0
+}
+
+/// Returns true if the chunk is safe to copy if unknown.
+const fn safe_to_copy(chunk: [u8; 4]) -> bool {
+    chunk[3] & 0b_00100000 != 0
+}
+
+use io::Write;
+use std::{io, ops};
+
+pub trait Chunk: ops::Deref<Target = [u8; 4]> {
+    const TYPE: [u8; 4];
+    const MAX_LENGTH: usize;
+
+    // -- Chunk type determination --
+
+    /// Returns true if the chunk is critical.
+    fn is_critical() -> bool {
+        is_critical(Self::TYPE)
+    }
+
+    /// Returns true if the chunk is private.
+    fn is_private() -> bool {
+        is_private(Self::TYPE)
+    }
+
+    /// Checks whether the reserved bit of the chunk name is set.
+    /// If it is set the chunk name is invalid.
+    fn reserved_set() -> bool {
+        reserved_set(Self::TYPE)
+    }
+
+    /// Returns true if the chunk is safe to copy if unknown.
+    fn safe_to_copy() -> bool {
+        safe_to_copy(Self::TYPE)
+    }
+
+    fn write<W: std::io::Write>(w: &mut W, data: &[u8]) -> io::Result<()> {
```

Comment:
This method makes no sense to me. Either the strongly typed `T: Chunk` knows exactly what kind of data it contains, or this should just be a method on a single monomorphic `struct Chunk([u8; 4])`.